packages/lisk-p2p/src/p2p.ts

```diff
@@ -166,50 +166,51 @@ export class P2P extends EventEmitter {
 		});
 	}
 
-	private async _loadListOfPeerListsFromSeeds(
-		seedList: ReadonlyArray<PeerInfo>,
-	): Promise<ReadonlyArray<ReadonlyArray<Peer>>> {
-		return Promise.all(
-			seedList.map(async (seedPeer: PeerInfo) => {
-				const peer = new Peer({
-					ipAddress: seedPeer.ipAddress,
-					wsPort: seedPeer.wsPort,
-					nodeStatus: this._nodeStatus,
-					height: seedPeer.height,
-					os: seedPeer.os,
-					version: seedPeer.version,
-				});
-				this._newPeers.add(peer.peerInfo);
-				/**
-				 * TODO LATER: For the LIP phase, we shouldn't add the seed peers to our
-				 * _peerPool and we should disconnect from them as soon as we've loaded their peer lists.
-				 */
-				this._peerPool.addPeer(peer);
-				// TODO LATER: For the LIP phase, the 'list' request will need to be renamed.
-				const seedNodePeerListResponse = await peer.request<void>({
-					procedure: 'list',
-				});
-				const peerListResponse = seedNodePeerListResponse.data as ProtocolPeerList;
-				// TODO ASAP: Validate the response before returning. Check that seedNodePeerListResponse.data.peers exists.
-
-				return peerListResponse.peers.map(
-					(peerObject: ProtocolPeerInfo) =>
-						new Peer({
-							ipAddress: peerObject.ip,
-							wsPort: peerObject.wsPort, // TODO ASAP: Add more properties
-							nodeStatus: this._nodeStatus,
-							height: seedPeer.height,
-							os: seedPeer.os,
-							version: seedPeer.version,
-						}),
-				);
-			}),
+	private _getPeerToConnect(): ReadonlyArray<Peer> {
+		const availablePeers = Array.from(this._newPeers).map(
+			(peerInfo: PeerInfo) => new Peer(peerInfo),
 		);
+		const peersToConnect = selectForConnection(availablePeers);
+
+		return peersToConnect;
+	}
+
+	private _connectToPeers(): void {
+		this._getPeerToConnect().map((peer: Peer) => {
+			peer.connect();
```

Comment:
shouldn't `peer.connect` be async?

Reply:
I think we are not waiting for the peer to connect, that's why we have 3 status in Peer.
```javascript
export enum ConnectionState {
	CONNECTING = 0,
	CONNECTED = 1,
	DISCONNECTED = 2,
}
```

@jondubois what do you think?

Reply:
Not necessary to wait for connect when adding to the `_newPeers` map.

For the `_triedPeers` map, I guess it depends on the LIP. Is the act of trying to connect to a peer enough to consider it a `_triedPeer` or is is necessary to wait for the connection to be successful?

It's possible to send RPCs and messages to a disconnected or 'connecting' peer (in that case it will buffer while it tries to establish the connection); so the peer is usable even if in a 'connecting' state. Once we've called 'connect()', the socket will always try to maintain the connection open and will automatically try to reconnect in case of network failure.

Reply:
The peer is usable in the 'connecting' state but because we're not physically connected yet, we won't receive any messages from a peer while it's in that state.

Reply:
In LIP `triedPeers`,

>  Containing peers to which there is or has been a successful outgoing connection. 

So that means we need to be sure that it was successfully connected and then we move it to `triedPeers`. At first if we try to connect to any peer in `newPeers`, just after first attempt if it fails to connect we remove it from `_newPeers`. In LIP, 

>  Peers in the collection newPeers are immediately deleted after a failed connection attempt as malicious peers can advertise invalid IP addresses.

So I think we need `async`, as we want to make sure we connect successfully with a peer

Reply:
But `SCClientSocket.connect(): void` doesn't return anything (`connect()` function in Peer simply attaches a outbound socket), we will have to explicitly get the status of a peer `this._outboundSocket.state`, after certain time so that we make sure it connects. Or, is there any other way to do it? @jondubois

Reply:
yes I think that we an make connect async and make it resolve once the `socket.on('connect', ...)` event triggers.